t_stress/test_flood.py

```diff
@@ -0,0 +1,141 @@
+           proxy_pass default;
+        }
+        """
+    }
+
+    @parameterize.expand(
+        [
+            param(
+                name="ping",
+            ),
+            param(
+                name="settings",
+            ),
+        ]
+    )
+    def test(self, name):
+        self.oops_ignore = ["ERROR"]
+        self.start_all_services()
+
+        hostname = tf_cfg.cfg.get("Tempesta", "hostname")
+        port = 443
+
+        context = ssl.create_default_context()
+        context.set_alpn_protocols(["h2"])
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        new_settings = dict()
+        new_settings[SettingCodes.MAX_CONCURRENT_STREAMS] = 100
+        ping_data = b"\x00\x01\x02\x03\x04\x05\x06\x07"
+
+        with socket.create_connection((hostname, port)) as sock:
+            with context.wrap_socket(sock, server_hostname="tempesta-tech.com") as ssock:
+                conn = h2.connection.H2Connection()
+                conn.initiate_connection()
+
+                for _ in range(1000_0000):
+                    if name == "ping":
+                        conn.ping(ping_data)
+                    else:
+                        conn.update_settings(new_settings)
+                    try:
+                        ssock.sendall(conn.data_to_send())
+                    except ssl.SSLEOFError:
+                        return
+        self.oops.find(
+            "ERROR: Too many control frames in send queue, closing connection",
+            cond=dmesg.amount_positive,
+        )
+
+    def test_reset_stream(self):
+        self.oops_ignore = ["ERROR"]
+        self.start_all_services()
+        client = self.get_client("deproxy")
+
+        hostname = tf_cfg.cfg.get("Tempesta", "hostname")
+        port = 443
+
+        context = ssl.create_default_context()
+        context.set_alpn_protocols(["h2"])
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+        headers = [
+            (":method", "GET"),
+            (":path", "/"),
+            (":authority", "tempesta-tech.com"),
+            (":scheme", "https"),
+        ]
+
+        with socket.create_connection((hostname, port)) as sock:
+            with context.wrap_socket(sock, server_hostname="tempesta-tech.com") as s:
+                c = h2.connection.H2Connection()
+                c.initiate_connection()
+                s.sendall(c.data_to_send())
+
+                for i in range(1, 100_0000, 2):
+                    hf = HeadersFrame(
+                        stream_id=i,
+                        data=client.h2_connection.encoder.encode(headers),
+                        flags=["END_HEADERS"],
+                    ).serialize()
+                    prio = PriorityFrame(stream_id=i, depends_on=i).serialize()
+                    try:
+                        s.sendall(hf + prio)
+                    except ssl.SSLEOFError:
+                        return
```

Comment:
```suggestion
        client.update_initial_settings()
        client.send_bytes(client.h2_connection.data_to_send())
        self.assertTrue(client.wait_for_ack_settings())
        client.readable = lambda : False  # disable socket for reading

        headers = [
            (":method", "GET"),
            (":path", "/"),
            (":authority", "tempesta-tech.com"),
            (":scheme", "https"),
        ]

        for i in range(1, 100_0000, 2):
            hf = HeadersFrame(
                stream_id=i,
                data=client.h2_connection.encoder.encode(headers),
                flags=["END_HEADERS"],
            ).serialize()
            prio = PriorityFrame(stream_id=i, depends_on=i).serialize()
            client.send_bytes(hf + prio, expect_response=False)
        self.assertTrue(client.wait_for_connection_close())
```

Reply:
ditto.

Reply:
for `test_reset_stream`: my example send `PRIORITY` frames in separate thread (deproxy manager thread) and wait for connection close in main thread. So my example requires ~10 sec because TempestaFW blocks connections until all frames are received. And I don't see OOM using `DeproxyClientH2` for local/remote setup. 
Your example requires ~90 sec for local setup and ~45 sec for remote setup.

Reply:
Sorry, I don't get it. I just reused your code directly. I did observe a rapid increase in client memory.
Anyway, why does `DeproxyClientH2` accumulate memory? It doesn't seem to be suitable for stress testing. Shortening the duration faster than the memory increment speed seems like a bad idea.

> because TempestaFW blocks connections until all frames are received. 

temepsta keeps sending control frame acks to the client, but the client doesn't read them, that's the reason for the OOM. So I don't understand what you mean, any further clarification?

Reply:
In your code
```python3
                for i in range(1, 100_0000, 2):
                    hf = HeadersFrame(
                        stream_id=i,
                        data=client.h2_connection.encoder.encode(headers),
                        flags=["END_HEADERS"],
                    ).serialize()
                    prio = PriorityFrame(stream_id=i, depends_on=i).serialize()
                    try:
                        s.sendall(hf + prio)
                    except ssl.SSLEOFError:
                        return
```
you send 100_0000 `HEADERS` + `PRIORITY` frames, but TempestaFW blocks this socket on the 1000 frame. TempestaFW will not receive other frames because socket will be closed. But you wait for sending all frames.

In my case, I just add frames to the buffer for sending, but don't wait for the frames to be sent.

You see memory consumption by the client because all frames store in buffers, but I think it is not a problem.

Reply:
> but TempestaFW blocks this socket on the 1000 frame. TempestaFW will not receive other frames because socket will be closed.

How did you observe this? It shouldn't be like this according to the code.
Tempesta has no limit on the processing of control frames, it continuously processes them and sends acks to the client without blocking.